### ISSUE TITLE
CloudWatch Log Event Subscriptions Deleted

### ISSUE THREAD

**Issue by lukespice** [serverless/serverless#5700]
# This is a Bug Report

## Description

We have a Serverless Lambda function that forwards logs from CloudWatch to Loggly similar to https://github.com/loggly/cloudwatch2loggly
After the update to 1.36.0 CloudWatch Logs triggers are deleted on deploy

## To reproduce
Deploy a function with 3 CloudWatch Log event triggers
```
functions:
  app:
    name: cloudwatch2loggly
    handler: src/index.handler
    events:
      - cloudwatchLog: '/aws/lambda/${self:provider.stage}-function1'
      - cloudwatchLog: '/aws/lambda/${self:provider.stage}-function2'
      - cloudwatchLog: '/aws/lambda/${self:provider.stage}-function3'
```
Run serverless deploy
`serverless deploy`
* Triggers are visible in the console and function is triggered as expected
![image](https://user-images.githubusercontent.com/797831/51149157-f5e33980-18b4-11e9-9833-22f1ac288ee7.png)

Run serverless deploy again (specify force as there are no code changes)
`serverless deploy --force`
* Triggers are deleted
![image](https://user-images.githubusercontent.com/797831/51149306-a5201080-18b5-11e9-937b-6254b4439481.png)

On each deployment I can see three calls to `DescribeSubscriptionFilters` in CloudTrail e.g.
```
    "requestParameters": {
        "logGroupName": "/aws/lambda/stage-function1"
    },
```
And on the deployment where the triggers are deleted we see three `DeleteSubscriptionFilter` in CloudTrail e.g.
```
"requestParameters": {
        "filterName": "cloudwatch2loggly-stage-AppLogsSubscriptionFilterCloudWatchLog1-XXXXXXXX",
        "logGroupName": "/aws/lambda/stage-function1"
    },
```

Drift detection also shows the subscriptions are deleted e.g. `AppLogsSubscriptionFilterCloudWatchLog1`
![image](https://user-images.githubusercontent.com/797831/51149885-1fea2b00-18b8-11e9-9370-1d6255303498.png)

Similar or dependent issues:
* Appears to be a result of the changes made in https://github.com/serverless/serverless/pull/5554

## Additional Data

* ***Serverless Framework Version you're using***: 1.36.1
* ***Operating System***: Windows 10
* ***Stack Trace***:
* ***Provider Error messages***:

**Comment by rdsedmundo** [comment]
Hey, I'm sorry my changes caused that. Going to try to reproduce the problem later.

**Comment by NishantRanjan** [comment]
Hi, please let me know the work around for this? Revert version back to 1.35.X?

**Comment by lukespice** [comment]
@NishantRanjan Yes, in our case we reverted to 1.35.1 and the problem was fixed.

**Comment by dschep** [comment]
Any progress @rdsedmundo?

**Comment by exoego** [comment]
> Triggers are deleted

Perhaps, this is related to #5833: CW Log Subscriptions do not appear in Lambda settings Triggers.

**Comment by rdsedmundo** [comment]
Sorry, I was really busy those last weeks. I'll take a look at this at the end of this week/weekend.

**Comment by lukespice** [comment]
@rdsedmundo Wondering if you ever get a chance to investigate? We're currently unable to upgrade due to this bug so missing out on new features and bug fixes.

**Comment by tbenade** [comment]
Hey @rdsedmundo thanks for the contributions to date. Any chance you could tell us if you are going to have a look at fixing this. This currently blocks people from moving to NodeJs10.x in later versions of serverless. If you could just confirm you will or wont then maybe someone else would be more willing to engage in the fix, thank you.

**Comment by rdsedmundo** [comment]
I took the time to investigate this today. I wasn't able to reproduce the issue following steps that OP mentioned, i.e to just add some triggers and then re-deploy with the same ones, without changing anything. It works here without errors.

But re #5817, i.e re-ordering under the same function, it's definitely something that I wasn't accounting for on my original PR. The problem is on how Serverless is naming the Logical IDs of the resources on the CloudFormation template. Example, let's say we have:

```
  hello:
    handler: handler.hello
    events:
      - cloudwatchLog: '/aws/lambda/subscription-test-1'
      - cloudwatchLog: '/aws/lambda/subscription-test-2'
      - cloudwatchLog: '/aws/lambda/subscription-test-3'
```

We'll generate the following resources names:

HelloLogsSubscriptionFilterCloudWatchLog1 (/aws/lambda/subscription-test-1)
HelloLogsSubscriptionFilterCloudWatchLog2 (/aws/lambda/subscription-test-2)
HelloLogsSubscriptionFilterCloudWatchLog3 (/aws/lambda/subscription-test-3)

If you then go and change the order on the yml file, for example to:

```
  hello:
    handler: handler.hello
    events:
      - cloudwatchLog: '/aws/lambda/subscription-test-2'
      - cloudwatchLog: '/aws/lambda/subscription-test-1'
      - cloudwatchLog: '/aws/lambda/subscription-test-3'
```

We'll generate this template for update:

HelloLogsSubscriptionFilterCloudWatchLog1 (/aws/lambda/subscription-test-2)
HelloLogsSubscriptionFilterCloudWatchLog2 (/aws/lambda/subscription-test-1)
HelloLogsSubscriptionFilterCloudWatchLog3 (/aws/lambda/subscription-test-3)

Pay attention to the collision that will happen, as now the HelloLogs...1 is pointing to `subscription-test-2` now and not `subscription-test-1` as before, so CloudFormation will interpret this as the same resource as before and will first try to update HelloLogs...1 to point to `subscription-test-2`, before deleting the original subscription we already had on it.

We probably will need to restructure how the event is added, perhaps adding a property called `uniqueName` where the user provides some identifier for the subscription filter and we append it at the end of the Logical ID. @pmuens @dschep

**Comment by sameer-s** [comment]
I was investigating this issue (the same one as the OP, where subscription filters would just disappear without being swapped around) today as it was causing me some issues. First I tried implementing the solution suggested above, where you use the name of the log group to tag each subscription filter rather than a serial number. While this did fix the issue in #5817, it didn't do anything about the one encountered here.

I did some more work, adding a few print statements, and noticed that in the following few lines:
~~~~
const oldDestinationArn = subscriptionFilter.destinationArn;
const filterName = subscriptionFilter.filterName;
const newDestinationArn = `arn:aws:lambda:${region}:${accountId}:function:${serviceName}-${stage}-${functionName}`;
~~~~
 the values of `oldDestinationArn` and `newDestinationArn` did not match. While the ending (after `function:`) of `oldDestinationArn` matched the actual ARN of the subscription log, and the `name` field in my `serverless.yml` file, the ending of `newDestinationArn` instead matched the name of the YAML object corresponding to my function.

In other words, the code which deleted the old log subscriptions only worked if the `name` field was not set in `serverless.yml` and the default name was used. This was easily fixed by passing in `functionObj` to the `fixLogGroupSubscriptionFilters` method and changing the last line of the snippet above instead to:
~~~
const newDestinationArn = `arn:aws:lambda:${region}:${accountId}:function:${functionObj.name}`;
~~~

This seemed to resolve the issue. If you'd like I'm happy to make a PR with my changes (or if you'd prefer, two, since I also have the changes corresponding to the other issue as described in my first paragraph).
